library/L1_Peripheral/lpc40xx/eeprom.hpp

```diff
@@ -0,0 +1,155 @@
+    constexpr float kNanosecond = 1E-9f;
+
+    // Initialzie EEPROM wait state register with number of wait states
+    // for each of its internal phases
+    // Phase 3 (15 ns)
+    eeprom_register->WSTATE |=
+        static_cast<uint8_t>((15 * kNanosecond * kSystemClock) + 1);
+    // Phase 2 (55 ns)
+    eeprom_register->WSTATE |=
+        (static_cast<uint8_t>((55 * kNanosecond * kSystemClock) + 1)) << 8;
+    // Phase 1 (35 ns)
+    eeprom_register->WSTATE |=
+        (static_cast<uint8_t>((35 * kNanosecond * kSystemClock) + 1)) << 16;
+
+    // Initialize EEPROM clock
+    eeprom_register->CLKDIV = static_cast<uint8_t>(kSystemClock / kEepromClk);
+  }
+
+  // The EEPROM can only be interfaced in 32-bit mode, which means that the
+  // address used must be a multiple of 4. The address entered will be cut
+  // down to a multiple of 4.
+  void Write(const uint8_t * wdata, uint32_t full_address,
+             size_t count) const override
+  {
+    constexpr uint8_t kMax6Bits = 0b11'1111;
+
+    // Writing a 32-bit number to an address that isn't a multiple of 4 will
+    // cause a hard fault, thus this safety is put into place to avoid that.
+    full_address &= 0b1111'1111'1100;
+
+    // Because the EEPROM uses 32-bit communication, write_data will be casted
+    // into a uint32_t *
+    const uint32_t * write_data = reinterpret_cast<const uint32_t *>(wdata);
+
+    // The first 6 bits in the address (MSB) dictate which page is being written
+    // to in the EEPROM, and the last 6 bits (LSB) dictate offset in the page
+    uint8_t page_count  = static_cast<uint8_t>(full_address >> 6);
+    uint8_t page_offset = full_address & 0b0000'0011'1111;
+
+    uint16_t address;
+
+    for (; count > 0; count -= 4)
+    {
+      // Page offset is incremented by 4 because we're writing 32 bits
+      constexpr uint8_t kOffsetInterval = 4;
+
+      address = static_cast<uint16_t>((page_count << 6) + page_offset);
+
+      eeprom_register->ADDR  = address;
+      eeprom_register->CMD   = kWrite32Bits;
+      eeprom_register->WDATA = write_data[0];
+
+      write_data++;
+      page_offset = static_cast<uint8_t>(page_offset + kOffsetInterval);
+
+      // If the 64 byte page buffer fills up, then it must be programmed to the
+      // EEPROM before continuing.
+      if (page_offset > kMax6Bits)
+      {
+        Program(address);
+        page_count++;
+        page_offset = 0;
+      }
+    }
+
+    // Program final information so that it's stored in the EEPROM
+    Program(address);
+  }
+
+  // The EEPROM is accessed through a 64 byte page buffer, and after it fills
+  // up, it must be programmed to the actual EEPROM. This function handles that.
+  void Program(uint32_t address) const
+  {
+    eeprom_register->ADDR  = address;
+    eeprom_register->CMD  = kEraseProgram;
+
+    // Poll status register bit to see when programming is finished
+    Wait((!(eeprom_register->INT_STATUS >> kProgramStatusBit)) & 1);

```

Comment:
You are going to need to turn this into a lambda for this to work.

Reply:
The Wait statement? What do you mean by work? It seems to be working right now.